docs/sources/flow/reference/components/loki.source.heroku.md

```diff
@@ -0,0 +1,136 @@
+`labels`                 | `map(string)`          | The labels to associate with each received Heroku record. | `{}`    | no
+`forward_to`             | `list(LogsReceiver)`   | List of receivers to send log entries to.                 |         | yes
+`relabel_rules`          | `RelabelRules`         | Relabeling rules to apply on log entries.                 | `{}`    | no
+
+The `relabel_rules` field can make use of the `rules` export value from a
+`loki.relabel` component to apply one or more relabeling rules to log entries
+before they're forwarded to the list of receivers in `forward_to`.
+
+## Blocks
+
+The following blocks are supported inside the definition of `loki.source.heroku`:
+
+Hierarchy | Name | Description | Required
+--------- | ---- | ----------- | --------
+listener | [listener] | Configures a listener for Heroku messages. | yes
+
+[listener]: #listener-block
+
+### listener block
+
+The `listener` block defines the listen address and port where the listener
+expects Heroku messages to be sent to.
+
+Name                     | Type          | Description | Default | Required
+------------------------ | ------------- | ----------- | ------- | --------
+`address`                | `string`      | The `<host>` address to listen to for heroku messages. | `0.0.0.0` | no
+`port`                   | `int`         | The `<port>` to listen to for heroku messages. | | yes
+
+Example Block
+
+```river
+listener {
+  address  = "127.0.0.1"
+  port = 8080
+}
+```
+
+## Labels
+
+The `labels` map is applied to every message that the component reads.
+
+The following internal labels are available for relabeling:
+- `__heroku_drain_host`
+- `__heroku_drain_app`
+- `__heroku_drain_proc`
+- `__heroku_drain_log_id`
+
+All url query params will be translated to `__heroku_drain_param_<name>`
+
+If the `X-Scope-OrgID` header is set it will be translated to `__tenant_id__`
+
+## Exported fields
+
+`loki.source.heroku` does not export any fields.
+
+## Component health
+
+`loki.source.heroku` is only reported as unhealthy if given an invalid
+configuration.
+
+## Debug information
+
+`loki.source.heroku` exposes some debug information per Heroku listener:
+* Whether the listener is currently running.
+* The listen address.
+
+## Debug metrics
+* `loki_source_heroku_drain_entries_total` (counter): Number of successful entries received by the Heroku target.
+* `loki_source_heroku_drain_parsing_errors_total` (counter): Number of parsing errors while receiving Heroku messages.
+
+## Example
+
+This example listens for Heroku messages over TCP in the specified port and forwards them to a `loki.write` component using the Heroku timestamp.
+
+```river
+loki.source.heroku "local" {
+  listener {
+    address  = "127.0.0.1"
+    port = 8080
+  }
+  use_incoming_timestamp = true
+  labels   = { component = "loki.source.heroku"} 
+  forward_to = [loki.write.local.receiver]
```

Comment:
Let's format these like `riverfmt` would do.

The easiest way would be to copy these over to a file and use the `agent fmt /path/to/file` command to get the formatted on stdout.

Reply:
Also, let's have the example listen on `0.0.0.0`, since that's most likely what users would have to do to be able to receive entries from Heroku.